File: NowegianDict/src/main/java/no/dict/services/SQLiteService.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package no.dict.services;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import no.dict.data.DictItem;
import no.dict.data.VisitHash;

/**
 * The sqlite service to load and save data into database
 * @author wxlfr_000
 */
public class SQLiteService {
	/**
	 * Interface of updating a data in database
	 * @author wxlfr_000
	 *
	 * @param <T>
	 */
	private static abstract class UpdateDatabase<T>{
		T data;
		public UpdateDatabase(T data){
			this.data = data;
		}
		public abstract List<String> getSQLSs(Statement stat) throws SQLException;

		public void update(){
			Connection connection = null;
			try {
				connection = getConnection();
				Statement stat = connection.createStatement();
				List<String> sqls = getSQLSs(stat);
				stat.close();
				connection.close();
				connection = SQLiteService.getConnection();
				connection.setAutoCommit(false);
				stat = connection.createStatement();
				for (String iter : sqls) {
					stat.addBatch(iter);
				}
				stat.executeBatch();
				connection.commit();
				connection.close();
			} catch (ClassNotFoundException e) {
			} catch (SQLException e) {
			} finally {
				if (connection != null) {
					try {
						connection.close();
					} catch (SQLException e) {
					}
				}
			}
		}
	}
	
	/**
	 * Interface of obtaining an object from a query record
	 * @author wxlfr_000
	 *
	 */
	static abstract class SQLResultHandler{
		public abstract <T> T handle(ResultSet result) throws SQLException;
	}

	private static final String DATABASE_URL = "jdbc:sqlite:dict.db";
	
	/**
	 * Handler of obtaining a dictionary item from a query record
	 */
	private static SQLResultHandler handler1 = new SQLResultHandler() {
		@SuppressWarnings("unchecked")
		@Override
		public DictItem handle(ResultSet result) throws SQLException {
			return getDictItem(result);
		}
	};
	
	private static final String JDBC_DRIVER = "org.sqlite.JDBC";

	private static boolean registered = false;
	
	/**
	 * Load the visit hash from database
	 * @param hash the visit hash
	 */
	public static void loadVisitHash(VisitHash hash){
		Connection connection = null;
		try {
			connection = getConnection();
			Statement stat = connection.createStatement();
			String sql = null;
			for (int index = 0; index < 2; ++index) {
				sql = "create table if not exists hash" + index + " (word text primary key not null);";
				stat.executeUpdate(sql);
			}
			for (int index = 0; index < 2; ++index) {
				// Query the old hosts and compare them with the current hosts
				sql = "select word from hash" + index;
				List<String> old_words = getList(stat, sql, "word");
				hash.addAll(index, old_words);
			}
		}catch(SQLException e){
		}catch(ClassNotFoundException e) { }
	}

	/**
	 * Load dictionary items from database
	 * @param words
	 */
	public static void loadWords(Map<String, DictItem> words){
		Connection connection = null;
		try {
			connection = getConnection();
			Statement stat = connection.createStatement();
			String sql = "select name from sqlite_master where type = 'table';";
			List<String> tables = getList(stat, sql, "name");
			for(String table : tables){
				if(table.length() != 1) continue;
				sql =  "select * from '" + table + "';";
				List<DictItem> list = getList(stat, sql, handler1);
				for(DictItem iter : list){
					words.put(iter.getWord() + "|" + iter.getExplain(), iter);
				}
			}
		}catch(SQLException e){
		}catch (ClassNotFoundException e) { }
	}

	/**
	 * Save dictionary items to database
	 * @param words
	 */
	public static void saveDictItem(Map<String, DictItem> words){
		UpdateDatabase<Map<String, DictItem>> update = new UpdateDatabase<Map<String, DictItem>>(words){
			public List<String> getSQLSs(Statement stat) throws SQLException{
				List<String> sqls = new ArrayList<String>();
				String sql = "";
				Map<String, DictItem> new_words = new HashMap<String, DictItem>();
				new_words.putAll(data);

				sql = "select name from sqlite_master where type = 'table';";
				List<String> tables = getList(stat, sql, "name");
				for(String table : tables){
					if(table.length() != 1) continue;
					sql =  "select * from '" + table + "';";
					List<DictItem> list = getList(stat, sql, handler1);
					for(DictItem iter : list){
						new_words.remove(iter.getWord() + "|" + iter.getExplain());
					}
				}

				for(DictItem iter : new_words.values()){
					String table = iter.getWord().substring(0, 1);
					if(!tables.contains(table)){
						sql = createTable(table);
						sqls.add(sql);
						tables.add(table);
					}
					sql = insert(iter);
					sqls.add(sql);
				}
				return sqls;
			}
		};
		update.update();
	}

	/**
	 * Save visit hash into database
	 * @param hash the visit hash
	 */
	public static void saveVisitHash(VisitHash hash){
		UpdateDatabase<VisitHash> update = new UpdateDatabase<VisitHash>(hash){
			public List<String> getSQLSs(Statement stat) throws SQLException{
				List<String> sqls = new ArrayList<String>();
				String sql = "";

				/**
				 * Create tables for visited words and unvisited words respectively.
				 * new_words used to store the current version of visited words and unvisited words.
				 */
				@SuppressWarnings("unchecked")
				Set<String>[] new_words = new HashSet[2];
				for (int index = 0; index < 2; ++index) {
					sql = "create table if not exists hash" + index + " (word text primary key not null);";
					new_words[index] = new HashSet<String>();
					new_words[index].addAll(data.get(index));
					sqls.add(sql);
				}

				/**
				 * Get the old version of visited words and unvisited words respectively.
				 * Update them according to comparison with the current version of visited words and unvisited words
				 * 1. Delete a word if a word is missed in the current version
				 * 2. Add a word if a word is new in the current version.
				 */
				for (int index = 0; index < 2; ++index) {
					// Query the old hosts and compare them with the current hosts
					sql = "select word from hash" + index;
					List<String> old_words = getList(stat, sql, "word");

					for (String word : old_words) {
						if (!new_words[index].contains(word)) {
							// If a host is kept, update its urls in database
							sqls.add("delete from hash" + index + " where word='" + toSQLTextValue(word) + "';");
						} 
						new_words[index].remove(word);
					}
				}
				for (int index = 0; index < 2; ++index) {
					for (String word : new_words[index]) {
						sql = "insert into hash" + index + " values ('" + toSQLTextValue(word) + "');";
						sqls.add(sql);
					}
				}
				return sqls;
			}
		};
		update.update();
	}
	/**
	 * Return the sql expression for creating a table to record words
	 * @param table
	 * @return the sql expression
	 */
	private static String createTable(String table){
		return "CREATE TABLE if not exists '" + table  
				+ "' (word TEXT NOT NULL,"
				+ " explain TEXT NOT NULL,"
				+ " sound TEXT,"
				+ " syllabel TEXT,"
				+ " clazz TEXT,"
				+ " format TEXT,"
				+ " composite TEXT,"
				+ " alternatives TEXT,"
				+ " comment TEXT,"
				+ " examples TEXT,"
				+ " phrases TEXT,"
				+ " error TEXT,"
				+ " audio BLOB,"
				+ " grammer TEXT, PRIMARY KEY(word, explain))";
	}
	private static Connection getConnection() throws ClassNotFoundException, SQLException {
		if (!registered) {
			Class.forName(JDBC_DRIVER);
			registered = true;
		}
		return DriverManager.getConnection(DATABASE_URL);
	}

	/**
	 * return a dictionary item from a record
	 * @param rs
	 * @return
	 */
	private static DictItem getDictItem(ResultSet rs) {
		DictItem result = new DictItem();
		try {
			result.setWord(rs.getString("word"));
			result.setExplain(rs.getString("explain"));
			result.setSound(rs.getString("sound"));
			result.setSyllabel(rs.getString("syllabel"));
			result.setClazz(rs.getString("clazz"));
			result.setFormat(rs.getString("format"));
			result.setComposite(rs.getString("composite"));
			result.setAlternative(rs.getString("alternatives"));
			result.setComment(rs.getString("comment"));
			result.setExamples(rs.getString("examples"));
			result.setPhrases(rs.getString("phrases"));
			result.setError(rs.getString("error"));
			result.setGrammer(rs.getString("grammer"));
		} catch (SQLException e) {
			return null;
		} 
		return result;
	}

	/**
	 * Return a list of element from query results 
	 * @param stat statement
	 * @param sql the sql expression to perform a query
	 * @param handler the handler to obtain an object from a query record
	 * @return a list of element
	 * @throws SQLException
	 */
	private static <T> List<T> getList(Statement stat, String sql, SQLResultHandler handler) throws SQLException{
		ResultSet result = null;
		List<T> list = new ArrayList<T>();
		try {
			result = stat.executeQuery(sql);
		} catch (SQLException e) {
			return list;
		}
		if (result != null)
			while (result.next()) {
				T item = handler.handle(result);
				if(item != null)
					list.add(item);
			}
		return list;
	}


//	public static ResultSet query(Statement stmt, String query) {
//		ResultSet rs = null;
//		try {
//			rs = stmt.executeQuery(query);
//		} catch (SQLException ex) {
//			Logger.getLogger(SQLiteService.class.getName()).log(Level.SEVERE, null, ex);
//			return null;
//		}
//		return rs;
//	}
	
	/**
	 * Get values of column from query result of sql expression
	 * @param stat
	 * @param sql
	 * @param column
	 * @return a list of strings for a certain column
	 * @throws SQLException
	 */
	private static List<String> getList(Statement stat, String sql, final String column) throws SQLException {
		SQLResultHandler handler = new SQLResultHandler() {

			@SuppressWarnings("unchecked")
			@Override
			public String handle(ResultSet result) throws SQLException {
				return result.getString(column);
			}
		};
		return getList(stat, sql, handler);
	}

	/**
	 * Return the sql expression to insert a record into a word table
	 * @param item
	 * @return the sql expression
	 */
	private static String insert(DictItem item){
		return "insert into '" + item.getWord().substring(0, 1) + "' values ('" 
				+ item.getWord().replaceAll("'", "''") + "','"
				+ item.getExplain().replaceAll("'", "''") + "','"
				+ item.getSound().replaceAll("'", "''") + "','"
				+ item.getSyllabel().replaceAll("'", "''") + "','"
				+ item.getClazz().replaceAll("'", "''") + "','"
				+ item.getFormat().replaceAll("'", "''") + "','"
				+ item.getComposite().replaceAll("'", "''") + "','"
				+ item.getAlternative().replaceAll("'", "''") + "','"
				+ item.getComment().replaceAll("'", "''") + "','"
				+ item.getExamples().toString().replaceAll("'", "''") + "','"
				+ item.getPhrases().toString().replaceAll("'", "''") + "','"
				+ item.getError().toString().replaceAll("'", "''") + "',"
				+ null + ",'"
				+ item.getGrammer().replaceAll("'", "''") + "')";
	}

	/**
	 * Change str to text value in sql expression
	 * @param str
	 * @return
	 */
	private static String toSQLTextValue(String str) {
		return str.replaceAll("'", "''");
	}
}

File: nordict/www/js/SQLiteService.js
var dictionary = null;
function initDB() {
    dictionary = window.sqlitePlugin.openDatabase({name: "dict.db", location: 'default'}, function(db){
        appConsole.print("Open database successfully!")
        createFavorite(db);
    }, function(e){
        appConsole.print("Error happens when opening database!")
        appConsole.print(e.message);
    });
}

var createFavorite = function(db){
    db.sqlBatch([
        "create table if not exists 'favorite' (word text not null, explain text not null, primary key(word, explain));",
        "create table if not exists 'history' (word text not null, explain text not null, primary key(word, explain));",
        "create table if not exists 'mydict' (word text not null, explain text not null, myexplain text not null, primary key(word, explain));"
        ],
        function(){
        appConsole.print("create tables 'favorite' and 'mydict' successfully");
    }, function(e){
        appConsole.print(e.message);
    });
}


function DictItem(word, explain) {
    this.word = word;
    this.explain = splitToArray(explain);
    if(this.explain.length > 0 && this.explain[0].indexOf(";") != -1){
        var array = this.explain;
        this.explain = array[0].split(";");
        for(var index = 1; index < array.length; ++index){
            this.explain.push(array[index]);
        }
    }
}

var splitToArray = function(str){
    var result = str.split("|");
    if(str.length == 0){
        result.length = 0;
    }
    return result;
}
var getField = function(row, field){
    var result = row[field];
    if(result == null)
        result = '';
    return result;
}
var getPhrases = function(phrases){
    var phraseArray = [];
    for(var iter in phrases){
        var cur = phrases[iter];
        var start = cur.indexOf('("');
        var key = cur.substring(0, start).trim();
        var end = cur.indexOf('")', start);
        if(end == -1) end = undefined;
        var value = cur.substring(start + 2, end);
        phraseArray.push([key, value]);
    }
    return phraseArray;
}
app.service('SQLiteService', function () {
    var getTableName = function (word) {
        if (word == null || word.length == 0)
            return null;
        var table = word.substr(0, 1);
        if(!!(table.match(/[a-z]/)))
            return table;
        return "other";
    };
    var perform = function (word, success, getSQL) {
        if (dictionary == null) return;
        var table = getTableName(word.toLowerCase());
        if (table == null) return;
        var sql = getSQL(table);
        if(sql.length > 0)
        dictionary.executeSql(sql, [], success, function (e) {
            if (appConsole != null) {
                appConsole.show();
                appConsole.clear();
                appConsole.print(e.message);
                appConsole.hide();
            };
        });
    };
    this.query = function (word, results, success) {
        results.length = 0;
        perform(word, function(res){
            for (var index = 0; index < res.rows.length; ++index) {
                results.push(getDictItemFromRecord(res.rows.item(index)));
            }
            if(!!success)
                success();
        }, function (table) {
            return "select * from '" + table + "' where word like '" + word + "%' and explain <> '' order by word limit 50;";
        });
    };
    this.query_exact = function (word, explain, results, success) {
        results.length = 0;
        perform(word, function(res){
            for (var index = 0; index < res.rows.length; ++index) {
                results.push(getDictItemFromRecord(res.rows.item(index)));
            }
            if(!!success)
                success();
        }, function (table) {
            return "select * from '" + table + "' where word='" + word + "' and explain like '" + explain + "%';";
        });
    };
    this.query_suggest = function (word, results, success) {
        results.length = 0;
        perform(word, function(res){
            for (var index = 0; index < res.rows.length; ++index) {
                var row = res.rows.item(index);
                results.push(new DictItem(getField(row, "word"), getField(row, "explain")));
            }
            if(!!success)
                success();
        }, function (table) {
            return "select word, explain from '" + table + "' where word like '" + word + "%' and explain <> '' order by word limit 50;";
        });
    };
    this.queryFavorite = function (word, results, success) {
        perform(word, function(res){
            var favorites = {};
            for(var index = 0; index < res.rows.length; ++index){
                var row = res.rows.item(index);
                favorites[getField(row, "word") + "|" + getField(row, "explain")] = true;
            }
            var favorite;
            for(var index in results){
                var item = results[index];
                item.favorite = !!(favorites[item.word + "|" + item.explain[0]]);
            }
            if(!!success)
                success();
        }, function (table) {
            return "select * from 'favorite' where word like '" + word + "%' order by word limit 50;";
        });
    };
    this.queryMyDict = function (word, results, success) {
        perform(word, function(res){
            var favorites = {};
            for(var index = 0; index < res.rows.length; ++index){
                var row = res.rows.item(index);
                favorites[getField(row, "word") + "|" + getField(row, "explain")] = getField(row, "myexplain");
            }
            var favorite;
            for(var index in results){
                var item = results[index];
                item.myDict = favorites[item.word + "|" + item.explain[0]];
                if(item.myDict == null) item.myDict = "";
            }
            if(!!success)
                success();
        }, function (table) {
            return "select * from 'mydict' where word like '" + word + "%' order by word limit 50;";
        });
    };
    this.deleteFavorite = function(item, success){
        perform(item.word, success, function (table) {
            return "delete from 'favorite' where word='" + item.word + "' and explain='" + item.explain[0] + "';";
        });
    };
    this.addFavorite = function(item, success){
        perform(item.word, success, function (table) {
            return "insert into 'favorite' values('" + item.word + "', '" + item.explain[0] + "');";
        });
    };
    this.saveToMyDict = function(item, success){
        perform(item.word, success, function (table) {
            if(item.myDict.length == 0){
                if(item.tempDict.length > 0){
                    return "insert into 'mydict' values('" + item.word + "', '" + item.explain[0] + "', '" + item.tempDict + "');";
                }
            }else{
                if(item.tempDict.length > 0) {
                    return "update 'mydict' set myexplain='" + item.tempDict + "' where word='" + item.word + "' and explain = '" + item.explain[0] + "';";
                } else {
                    return "delete from 'mydict' where word='" + item.word + "' and explain= '" + item.explain[0] + "';"
                }
            }
            return "";
        });
    };
});

getDictItemFromRecord = function(row)
{
    try {
        var result = new DictItem(getField(row, "word"), getField(row, "explain"));
        result.sound = getField(row, "sound");
        result.syllabel = getField(row, "syllabel");
        result.clazz = getField(row, "clazz");
        result.format = getField(row, "format").replace(' ]', ']');
        result.composite = splitToArray(getField(row, "composite"));
        result.composite.forEach(function (iter, index, arr) {
            var i = iter.indexOf(') ');
            if (i != -1)
                arr[index] = iter.substring(i + 2);
        });
        result.alternatives = getField(row, "alternatives");
        result.comment = getField(row, "comment");
        result.examples = splitToArray(getField(row, "examples"));
        var phrases = splitToArray(getField(row, "phrases"));
        result.phrases = getPhrases(phrases);
        return result;
    }catch (e){
        appConsole.print(e.message);
    }
};

